src/fact_samplers.jl

```diff
@@ -0,0 +1,186 @@
+# implementation of factorised samplers
+using DataStructures
+using Statistics
+using SparseArrays
+using LinearAlgebra
+
+"""
+    neighbours(G::Vector{<:Pair}, i) = G[i].second
+
+Return extended neighbourhood of `i` including `i`.
+`G`: graphs of neightbourhoods
+"""
+neighbours(G::Vector{<:Pair}, i) = G[i].second
+
+#need refreshments
+hasrefresh(::FactBoomerang) = true
+hasrefresh(::ZigZag) = false
+
+
+normsq(x::Real) = abs2(x)
+normsq(x) = dot(x,x)
+
+isZigZagTrace(::ZigZag) = true
+isZigZagTrace(::FactBoomerang) = false
+
+"""
+    λ(∇ϕ, i, x, θ, Z::ZigZag)
+`i`th Poisson rate of the `ZigZag` sampler
+"""
+function λ(∇ϕ, i, x, θ, Z::ZigZag)
+    pos(∇ϕ(x, i)*θ[i])
+end
+
+"""
+    λ(∇ϕ, i, x, θ, Z::FactBoomerang)
+`i`th Poisson rate of the `FactBoomerang` sampler
+"""
+function λ(∇ϕ, i, x, θ, B::FactBoomerang)
+    pos((∇ϕ(x, i) - (x[i] - B.μ[i]))*θ[i])
+end
+
+
+"""
+    ab(G, i, x, θ, c, Z::ZigZag)
+
+Returns the constant term `a` and linear term `b` when computing the Poisson times
+from the upper upper bounding rates λᵢ(t) = max(a + b*t)^2. The factors `a` and `b`
+can be function of the current position `x`, velocity `θ`, tuning parameter `c` and
+the Graph `G`
+"""
+function ab(G, i, x, θ, c, Z::ZigZag)
+    a = c[i] + θ[i]*(dot(Z.Γ[:, i], x)  - dot(Z.Γ[:, i], Z.μ))
+    b = θ[i]*dot(Z.Γ[:, i], θ)
+    a, b
+end
+
+"""
+    ab(G, i, x, θ, c, Z::FactBoomerang)
+
+Returns the constant term `a` and linear term `b` when computing the Poisson times
+from the upper upper bounding rates λᵢ(t) = max(a + b*t)^2. The factors `a` and `b`
+can be function of the current position `x`, velocity `θ`, tuning parameter `c` and
+the Graph `G`
+"""
+function ab(G, i, x, θ, c, Z::FactBoomerang)
+    nhd = neighbours(G, i)
+    a = c[i]*sqrt(normsq(x[nhd] - Z.μ[nhd]) + normsq(θ[nhd]))
+    b = 0.0
+    a, b
+end
+
+"""
+    λ_bar(G, i, x, θ, c, Z::ZigZag)
+
+Computes the bounding rate `λ_bar` at position `x` and velocity `θ`.
+"""
+λ_bar(G, i, x, θ, c, Z::ZigZag) = pos(ab(G, i, x, θ, c, Z::ZigZag)[1])
+
+"""
+    λ_bar(G, i, x, θ, c, Z::FactBoomerang)
+
+Computes the bounding rate `λ_bar` at position `x` and velocity `θ`.
+"""
+λ_bar(G, i, x, θ, c, Z::FactBoomerang) = pos(ab(G, i, x, θ, c, Z)[1])
+
+
+function event(i, t, x, θ, Z::ZigZag)
+     t, i, x[i], θ[i]
+end
+
+function event(i, t, x, θ, Z::FactBoomerang)
+    t, i, x, θ
```

Comment:
Not `x[i]` and `θ[i]`?

Reply:
Ah! You are right! `x[i]` and `θ[i]` are sufficient.